jetty-server/src/test/java/org/eclipse/jetty/server/BlockingTest.java

```diff
@@ -0,0 +1,504 @@
+    private ContextHandler context;
+
+    @BeforeEach
+    void setUp()
+    {
+        server = new Server();
+        connector = new ServerConnector(server);
+        connector.setPort(0);
+        server.addConnector(connector);
+
+        context = new ContextHandler("/ctx");
+
+        HandlerList handlers = new HandlerList();
+        handlers.setHandlers(new Handler[]{context, new DefaultHandler()});
+        server.setHandler(handlers);
+    }
+
+    @AfterEach
+    void tearDown() throws Exception
+    {
+        server.stop();
+    }
+
+    @Test
+    public void testBlockingReadThenNormalComplete() throws Exception
+    {
+        CountDownLatch started = new CountDownLatch(1);
+        CountDownLatch stopped = new CountDownLatch(1);
+        AtomicReference<Throwable> readException = new AtomicReference<>();
+        AbstractHandler handler = new AbstractHandler()
+        {
+            @Override
+            public void handle(String target, Request baseRequest, HttpServletRequest request, HttpServletResponse response) throws IOException, ServletException
+            {
+                baseRequest.setHandled(true);
+                new Thread(() ->
+                {
+                    try
+                    {
+                        int b = baseRequest.getHttpInput().read();
+                        if (b == '1')
+                        {
+                            started.countDown();
+                            if (baseRequest.getHttpInput().read() > Integer.MIN_VALUE)
+                                throw new IllegalStateException();
+                        }
+                    }
+                    catch (Throwable t)
+                    {
+                        readException.set(t);
+                        stopped.countDown();
+                    }
+                }).start();
+
+                try
+                {
+                    // wait for thread to start and read first byte
+                    started.await(10, TimeUnit.SECONDS);
+                    // give it time to block on second byte
+                    Thread.sleep(1000);
+                }
+                catch (Throwable e)
+                {
+                    throw new ServletException(e);
+                }
+
+                response.setStatus(200);
+                response.setContentType("text/plain");
+                response.getOutputStream().print("OK\r\n");
+            }
+        };
+        context.setHandler(handler);
+        server.start();
+
+        StringBuilder request = new StringBuilder();
+        request.append("POST /ctx/path/info HTTP/1.1\r\n")
+            .append("Host: localhost\r\n")
+            .append("Content-Type: test/data\r\n")
+            .append("Content-Length: 2\r\n")
+            .append("\r\n")
+            .append("1");
+
+        int port = connector.getLocalPort();
+        try (Socket socket = new Socket("localhost", port))
+        {
+            socket.setSoTimeout(1000000);
```

Comment:
No point in setting a timeout this large (17 mins). Don't set it, or make it a few seconds.